README.md
# Ice Boar

![Ice Boar](ice-boar/src/main/javadoc/resources/ice-boar-logo.png)

Helper for running a Java standalone application using Java Web Start (JNLP) with a custom version of JRE.

![build status](https://img.shields.io/badge/build-status%20=%3E-yellow.svg) [![Build Status](https://travis-ci.org/Roche/IceBoar.svg?branch=master)](https://travis-ci.org/Roche/IceBoar) [![codecov.io](https://codecov.io/github/Roche/IceBoar/coverage.svg?branch=master)](https://codecov.io/github/Roche/IceBoar?branch=master) 
[![Apache License Version 2.0](https://img.shields.io/badge/license-Apache%202.0-green.svg)](https://github.com/Roche/IceBoar/blob/master/LICENSE.txt)

![latest stable](https://img.shields.io/badge/latest-stable%20=%3E-yellow.svg) [ ![download bintray]
(https://img.shields.io/bintray/v/stachnim/IceBoar/com.roche.svg)](https://bintray.com/stachnim/IceBoar/com.roche/view) [ 
![download Maven Central](https://img.shields.io/maven-central/v/com.roche/ice-boar.svg)](http://search.maven.org/#search%7Cga%7C1%7Cg%3A%22com.roche%22%20AND%20a%3A%22ice-boar%22)

![latest snapshots](https://img.shields.io/badge/latest-snapshot%20=%3E-yellow.svg) [ ![download bintray]
(https://img.shields.io/badge/download-bintray-green.svg)](https://oss.jfrog.org/artifactory/webapp/#/artifacts/browse/tree/General/oss-snapshot-local/com/roche/ice-boar)

## Live demo: [roche.github.io/IceBoar](http://roche.github.io/IceBoar/)

Ice Boar is a small app started by Java Web Start (using JNLP file), with does the following:

* checks installed Java Version,
* downloads a JRE from the custom location (ZIP file) and unpacks it,
* downloads target JAR files and executes it on dedicated JVM (using java command instead of javaws or jp2launcher).

This helps to avoid typical problems with Java Web Start technology:

* your app is running on dedicated JVM prepared by you - no incompatibility issues anymore
* no super user rights needed to install JVM
* no more problems with web start sandbox (security, cache)
* you need to sign only IceBoar JAR file
* and others...       

# Release Notes
See [Release Notes](RELEASE-NOTES.md)

# What is it good for?
Good for project which uses Java Web Start and has problems with this technology.

# What is it not good for?
If you begin a new project and think about Java Web Start or Ice Boar don't do that if you don't really need it! Java
 Web Start (and Ice Boar based on it) will be not longer supported in Chrome. Google plan to remove [NPAPI]
 (https://en.wikipedia.org/wiki/NPAPI) support for Chrome. See more details here: [The Final Countdown for NPAPI]
 (http://blog.chromium.org/2014/11/the-final-countdown-for-npapi.html).

#How it Works?
The best way to describe how it works is below sequence diagram.

![Ice Boar Sequence Diagram](ice-boar/src/main/javadoc/resources/ice-boar-sequence-diagram.png)

User clicks a link in Web browser and downloads a JNLP file. By clicking downloaded file a Web browser downloads a Java
Runtime Environment (JRE), installs it, downloads this ice-boar library and runs it. At this point a standard Java
Web Start technology is used.

# How to use Ice Boar?
* You need to add ice-boar library to your project and make it downloadable. It is available via [Maven Central]
(http://search.maven.org/#search%7Cga%7C1%7Cice-boar):

```xml
<dependency>
    <groupId>com.roche</groupId>
    <artifactId>ice-boar</artifactId>
    <version>0.5</version>
    <classifier>jar-with-dependencies</classifier>
</dependency>
```

I suggest to use jar-with-dependencies. In other case you need to define additional Ice Boar dependencies in your 
JNLP file, using `jar` tag.

* You need to create a JNLP file and make it also downloadable. 
You can use [webstart-maven-plugin](http://www.mojohaus.org/webstart/webstart-maven-plugin/) for both or make it manually. 
To learn how it can be done please check [ice-boar-samples/ice-boar-demo](ice-boar-samples/ice-boar-demo) module.   

* Another code examples you can find on [live demo page](http://roche.github.io/IceBoar/). A source code of this page
 is available in [ice-boar-samples/ice-boar-demo-gh-pages/](ice-boar-samples/ice-boar-demo-gh-pages/).

## A minimum example of JNLP file
```xml
<?xml version="1.0" encoding="utf-8"?>

<!-- Define a sandbox by specifying codebase attribute. Based on this URL an absolute path for JARs will be created. -->
<!-- If you use relative paths, you can specify here attribute href (eg. href="http://SERVER:PORT/FILE.jnlp"). In -->
<!-- this case JNLP file will be downloaded by Web Start second time after start. -->
<jnlp spec="1.0+" codebase="http://SERVER:PORT">

<!-- Basic information. -->
<information>
   <title>Hello World Example Ice Boar</title>
   <vendor>Roche.com</vendor>
   <description>Hello World Description</description>
   <offline-allowed/>
</information>

<!-- ice-bore-core needs all permissions to work properly. The same configuration of permissions need be specified -->
<!-- in META-INF/MANIFEST.MF in JAR file. -->
<security>
   <all-permissions/>
</security>

<!-- Common definitions for all Environment -->
<resources>

   <!-- Defines a version of Java to be used to start IceBoar. It is preferred to use the same version as the target -->
   <!-- application uses. IceBoar is compiled for Java 1.5 -->
   <j2se version="1.5+" href="http://java.sun.com/products/autodl/j2se"/>

   <!-- Defines where JAR file with IceBoar class is located. In a corporate environment usually this JAR needs to --> 
   <!-- be signed to be executed with all permissions. -->
   <jar href="http://SERVER:PORT/PATH/ice-boar-VERSION-jar-with-dependencies.jar"/>

   <!-- Settings section. By using properties with "jnlp." prefix you can pass properties to your target --> 
   <!-- application. Alternatively you can use properties predefined in Java (e.g. path.separator, 
   <!-- deployment.javaws.concurrentDownloads, user.language). Other properties are ignored. Please check --> 
   <!-- GlobalSettings class javadoc for more useful properties. -->
   <property name="jnlp.IceBoar.frameTitle" value="Roche..."/>

   <!-- When set to true a window with debug information is displayed. Default is false. -->
   <property name="jnlp.IceBoar.showDebug" value="true"/>

   <!-- Defines a version of Java to execute the target application. -->
   <property name="jnlp.IceBoar.targetJavaVersion" value="1.7.0_04"/>

   <!-- Defines an entry point class to your / target application. -->
   <property name="jnlp.IceBoar.main-class" value="com.roche.iceboar.helloworldswing.HelloWorld"/>
 
   <!-- Defines a target application JAR. If you do not put all dependencies in target application JAR you can --> 
   <!-- specify more dependencies by providing more properties with next numbers as the prefix, e.g. --> 
   <!-- jnlp.IceBoar.jar.1, jnlp.IceBoar.2,... -->
   <property name="jnlp.IceBoar.jar.0" value="http://SERVER:PORT/PATH/hello-world-swing.jar"/>

   <!-- You can specify your custom properties by using prefix "jnlp.". All properties that do not start with --> 
   <!-- "jnlp.IceBoar." will be pass to the target application. -->
   <property name="jnlp.my.super.property" value="some value"/>
</resources>

<!-- Specifies properties specific to a given OS and architecture. -->
<resources os="Windows" arch="amd64">
   <property name="jnlp.IceBoar.targetJavaURL" value="http://SERVER:PORT/PATH/jre-1.7.0_04-win-x64.zip"/>
</resources>

<resources os="Mac OS X" arch="x86_64">
   <property name="jnlp.IceBoar.targetJavaURL" value="http://SERVER:PORT/PATH/jre-1.7.0_04-macosx-x64.zip"/>
</resources>

<!-- Definition of Entry Point to Ice Boar, witch wills run class defined in jnlp.IceBoar.main-class. -->
<application-desc main-class="com.roche.iceboar.IceBoar">
   <!-- Command line arguments for target application, if you need. Ice Boar doesn't require any arguments. -->
   <argument>arg1</argument>
   <argument>arg2</argument>
</application-desc>

</jnlp>
```

For more configuration options see GlobalSettings class javadoc.

## Sign Ice Boar JAR file
To use Ice Boar in production you need/should sign Ice Boar with your certificate, e.g.:

`jarsigner -keystore keyStore.jks -storepass <password> -tsa http://timestamp.digicert.com ice-boar-<version>-jar-with-dependencies.jar <alias>`

I prefer to use JAR with all dependencies. If you use Maven you can sign using 
[webstart-maven-plugin](http://www.mojohaus.org/webstart/webstart-maven-plugin/) or 
[maven-jarsigner-plugin](https://maven.apache.org/plugins/maven-jarsigner-plugin/).

# How to build a project?
To build IceBoar project from sources please use maven and execute `mvn clean install` on [ice-boar](ice-boar). This 
will produce 2 JAR's (one regular and second with all dependencies). You can use them in your project. For examples for
usage please see [ice-boar-samples/ice-boar-demo-gh-pages](ice-boar-samples/ice-boar-demo-gh-pages) or 
[ice-boar-samples/ice-boar-demo](ice-boar-samples/ice-boar-demo).

If you would to build all sub-projects please execute:
* `mvn clean install` in [ice-boar](ice-boar) directory
* `mvn clean install` in root (this)


# How to contribute?

Please read [CONTRIBUTING](CONTRIBUTING.md)

# License

This project is licensed under the terms of the Apache License Version 2.0. This license applies to all files
distributed in this project. You can see the whole license here: [LICENSE.txt](LICENSE.txt).

This project contains also 2 files (MessageConsole.java and LimitLinesDocumentListener.java) from 
[tips4java.wordpress.com](https://tips4java.wordpress.com) site. This files are free to use based on 
[about](https://tips4java.wordpress.com/about/) site: 
> You are free to use and/or modify and/or distribute any or all code posted on the Java Tips Weblog without restriction. A credit in the code comments would be nice, but not in any way mandatory.

To avoid any vagueness there is screenshot of this site (access: 02.12.2015)

[![tips4java about](license/tips4java.wordpress.com-about_small.png)](license/tips4java.wordpress.com-about.png)

# FAQ (Frequently Asked Questions)

## Q: Why IceBoar use System.out and System.err instead of SLF4J or others logging frameworks? 

The main goal of IceBoar was a small size of final JAR file. This library contains only necessary dependencies (I 
can probably avoid commons-lang3). The second reason to avoid logging frameworks is a short life cycle of IceBoar. 
For big projects it works maximum 10 Minutes and I don't need log rotation, patterns, and all this stuff from logging 
frameworks.

## Q: Why if I would to build a whole project I need to execute `mvn clean install` in two directories?

The [ice-boar](ice-boar) module is not included in parent pom.xml. Therefore I don't need to release parent pom in to
 Maven Central repository - It's not a trash.

# Authors
This project was initially developed by Marcin Stachniuk (Roche ADMD) to address the issues with Web Start reported by 
Michael Vogel from Genentech Bioinformatics & Computational Biology Department.

# Acknowledgments

Thanks Rafał Udziela (Roche ADMD) for coax me into writing this project.

Thanks code reviewers: Rafał Udziela (Roche ADMD) and Damian Mierzwiński (Roche ADMD).

Thanks for good usage feedback and first tests: Łukasz Migdałek (Roche ADMD).

Thanks for preparing logo: Paweł Opozda (Roche ADMD).

RELEASE-NOTES.md
## **1.0** (26.04.2016)
 * Run target application on pre-installed Java version if version match 
 (PR [#5](https://github.com/Roche/IceBoar/pull/5)) Thanks to koziolek for commitment
 * Add possibility to force use target JVM Version

## **0.7** (19.04.2016)
 * Add splash screen during downloading JRE and JAR files (PR [#4](https://github.com/Roche/IceBoar/pull/4)) Thanks 
 to Patryk Zieliński and Rafał Piotrowski for commitment
 * Add possibility to hide frame border 
 * Fix bug running a target application on Mac, remove blank command line arguments
 * Create gh-pages (live demo): [roche.github.io/IceBoar](http://roche.github.io/IceBoar/)
 * Add ice-boar-demo-gh-pages project and automate building live demo after each build
 * Change JRE version in examples
 * Always build Javadoc and Jar with source code during `mvn install` in ice-boar   

## **0.6** (19.02.2016)
 * Add Ice Boar Icon
 * Possibility to change default icon in Ice Boar

## **0.5** (13.01.2016)
 * Ice Boar available on Maven Central

## **0.5** (22.12.2015)
 * First public version of IceBoar (available on GitHub)



ice-boar/src/main/java/com/roche/iceboar/settings/GlobalSettings.java
/*
 * ****************************************************************************
 *  Copyright © 2015 Hoffmann-La Roche
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 * ****************************************************************************
 */

package com.roche.iceboar.settings;

import com.roche.iceboar.IceBoarException;
import com.roche.iceboar.cachestorage.CacheStatus;
import com.roche.iceboar.cachestorage.StatusInfo;
import com.roche.iceboar.downloader.FileUtilsFacade;
import com.roche.iceboar.runner.JVMVersionMatcher;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.roche.iceboar.downloader.FileUtilsFacade.extractFilenameFromURL;
import static org.apache.commons.lang3.StringUtils.isNotBlank;
import static org.apache.commons.lang3.StringUtils.join;

/**
 * This class defines properties that can be set in JNLP file to configure behaviour of Ice Boar application,
 * ex. <br>
 * <pre class="code"><code class="xml">{@code <resources os="Windows" arch="x86">
 * <property name="jnlp.IceBoar.targetJavaVersion" value="1.6.0_15"/>
 * </resources>}</code></pre>
 */
public class GlobalSettings {

    /**
     * <tt>{@value #JNLP_SHOW_DEBUG}</tt><br>
     * If this property is set to <tt>true</tt> a debug JFrame is shown. It is not recommended to use it on
     * production environment. Default set to <tt>false</tt>.
     */
    public static final String JNLP_SHOW_DEBUG = "jnlp.IceBoar.showDebug";

    /**
     * <tt>{@value #JNLP_FRAME_TITLE}</tt><br>
     * A title of Progress frame. Default set to: {@link GlobalSettingsFactory#DEFAULT_FRAME_TITLE}.
     */
    public static final String JNLP_FRAME_TITLE = "jnlp.IceBoar.frameTitle";

    /**
     * <tt>{@value #JNLP_TARGET_JAVA_VERSION}</tt><br>
     * This property defines a Java (JRE) Version. In plain JNLP (when you do not use ice-boar) this is
     * configured by <tt>&lt;j2se&gt;</tt> tag, key: version, ex:<br>
     * <pre class="code"><code class="xml">
     * {@code
     * <resources os="Windows" arch="amd64">
     * <j2se version="1.6+" ...>}
     * </code></pre>
     * It is highly recommend to set this to version the same like in &lt;j2se&gt; tag - if user doesn't have
     * Java installed in defined or newest version, then  plain Java Web Start will download and install JRE in a
     * version as specified (or newest).
     */
    public static final String JNLP_TARGET_JAVA_VERSION = "jnlp.IceBoar.targetJavaVersion";

    /**
     * <tt>{@value #JNLP_TARGET_JAVA_URL}</tt><br>
     * This property defines a location of downloadable JRE ZIP file. The downloaded JRE must be in the same version
     * as defined in {@link #JNLP_TARGET_JAVA_VERSION}.
     */
    public static final String JNLP_TARGET_JAVA_URL = "jnlp.IceBoar.targetJavaURL";

    /**
     * <tt>{@value #JNLP_JARS_PREFIX}</tt><br>
     * This prefix is used to define properties that specify libraries needed. At least one dependency needs to be
     * specified. In clear JNLP this properties correspond to:<br>
     * <pre class="code"><code class="xml">
     * {@code
     * <jar href="guava-XX.X.jar"/>
     * <jar href="slf4j-api-X.X.X.jar"/>}
     * </code></pre>
     * You can't duplicate the property name. If you need to specify more than one entry, please create more
     * properties that start from the same prefix e.g.
     * <pre class="code"><code class="xml">
     * {@code
     * <property name="jnlp.IceBoar.jar.0" value="guava-XX.X.jar"/>
     * <property name="jnlp.IceBoar.jar.1" value="slf4j-api-X.X.X.jar"/>}
     * </code></pre>
     */
    public static final String JNLP_JARS_PREFIX = "jnlp.IceBoar.jar.";

    /**
     * <tt>{@value #JNLP_MAIN_CLASS}</tt><br>
     * This property defines the target main class. It is a actual code to be executed on a client machine and
     * corresponds to the value of main-class property specified in <tt>&lt;application-desc&gt;</tt> tag, e.g.:<br>
     * <pre class="code"><code class="xml">
     * {@code
     * <application-desc main-class="com.roche.MainClass">}
     * </code></pre>
     */
    public static final String JNLP_MAIN_CLASS = "jnlp.IceBoar.main-class";

    /**
     * <tt>{@value #JNLP_INITIAL_HEAP_SIZE}</tt>
     * This property specifies an initial size of Java heap size to be used while executing the target application.
     * It corresponds to initial-heap-size value in <tt>&lt;j2se&gt;</tt> tag, e.g.:<br>
     * <pre class="code"><code class="xml">
     * {@code
     * <j2se href="..." initial-heap-size="128m"/>}
     * </code></pre>
     */
    public static final String JNLP_INITIAL_HEAP_SIZE = "jnlp.IceBoar.initial-heap-size";

    /**
     * <tt>{@value #JNLP_MAX_HEAP_SIZE}</tt>
     * This property define a max size of Java heap for target application.
     * It correspond to max-heap-size value in <tt>&lt;j2se&gt;</tt> tag, e.g.:<br>
     * <pre class="code"><code class="xml">
     * {@code
     * <j2se href="..." max-heap-size="1024m"/>}
     * </code></pre>
     */
    public static final String JNLP_MAX_HEAP_SIZE = "jnlp.IceBoar.max-heap-size";

    /**
     * <tt>{@value #JNLP_JAVA_VM_ARGS}</tt>
     * This property defines an additional set of standard and non-standard virtual machine arguments for the target
     * application. It corresponds to java-vm-args value in <tt>&lt;j2se&gt;</tt> tag, e.g.:<br>
     * <pre class="code"><code class="xml">
     * {@code
     * <j2se href="..." java-vm-args="-XX:+UseParallelGC"/>}
     * </code></pre>
     */
    public static final String JNLP_JAVA_VM_ARGS = "jnlp.IceBoar.java-vm-args";

    /**
     * <tt>{@value #JNLP_CLOSE_ON_END}</tt>
     * This property defines a behaviour after finishing staring a target application. When set on true, all Ice Boar
     * windows will be close, when false windows will be NOT close. The second approach is useful for debugging.
     * Default is set to true.
     */
    public static final String JNLP_CLOSE_ON_END = "jnlp.IceBoar.close-on-end";

    /**
     * <tt>{@value #JNLP_ICONS_PREFIX}</tt><br>
     * This prefix is used to define properties that specify icons for JFrame with progressbar. In clear JNLP this
     * properties correspond to:<br>
     * <pre class="code"><code class="xml">
     * {@code
     * <icon href="images/image.jpg"/>}
     * </code></pre>
     * You can't duplicate the property name. If you need to specify more than one entry (e.g. for different sizes
     * like 16x16, 32x32, 64x64, 128x128), please create more properties that start from the same prefix e.g.
     * <pre class="code"><code class="xml">
     * {@code
     * <property name="jnlp.IceBoar.icon.0" value="images/icon-128x128.png"/>
     * <property name="jnlp.IceBoar.icon.1" value="images/icon-64x64.png"/>}
     * </code></pre>
     *
     * @since 0.6
     */
    public static final String JNLP_ICONS_PREFIX = "jnlp.IceBoar.icon.";

    /**
     * <tt>{@value #JNLP_SPLASH_SCREEN}</tt><br>
     * Define your custom splash screen. If file will be not found a default frame will be showed.
     *
     * @since 0.7
     */
    public static final String JNLP_SPLASH_SCREEN = "jnlp.IceBoar.splash";

    /**
     * <tt>{@value #JNLP_SPLASH_HIDE_FRAME_BORDER}</tt><br>
     * Hide a frame border. A JFrame will be undecorated. Default is set to false.
     *
     * @since 0.7
     */
    public static final String JNLP_SPLASH_HIDE_FRAME_BORDER = "jnlp.IceBoar.hide-frame-border";

    /**
     * <tt>{@value #JNLP_ALWAYS_RUN_ON_TARGET_JVM}</tt><br>
     * Enforce always download JVM even if current JVM match to required target version. Default false.
     *
     * @since 0.8
     */
    public static final String JNLP_ALWAYS_RUN_ON_TARGET_JVM = "jnlp.IceBoar.alwaysRunOnTargetJVM";

    private JVMVersionMatcher versionMatcher = new JVMVersionMatcher();

    private List<String> applicationArguments;
    private long jvmStartTime;
    private boolean showDebug;
    private String frameTitle;
    private String targetJavaVersion;
    private String currentJavaVersion;
    private String targetJavaURL;
    private String tempDirectory;
    private List<String> jarURLs;
    private String mainClass;
    private List<String> allPropertiesForTarget;
    private String operationSystemName;
    private String pathSeparator;
    private String initialHeapSize;
    private String maxHeapSize;
    private String javaVmArgs;
    private String cachePath;
    private CacheStatus cacheStatus;
    private boolean closeOnEnd;
    private List<String> icons;
    private String customSplashImage;
    private boolean hideFrameBorder;
    private boolean alwaysRunOnTargetJVM;
    private String currentJavaCommand;


    /**
     * Arguments that come from JNLP file and should be propagated to the destination JAR.
     */
    private GlobalSettings(String[] applicationArguments) {
        if (applicationArguments == null || applicationArguments.length == 0) {
            this.applicationArguments = new ArrayList<String>();
        } else {
            this.applicationArguments = Arrays.asList(applicationArguments);
        }
    }

    public static Builder builder() {
        return new Builder();
    }

    public boolean isShowDebug() {
        return showDebug;
    }

    public String getFrameTitle() {
        return frameTitle;
    }

    public String getTargetJavaVersion() {
        return targetJavaVersion;
    }

    public String getCurrentJavaVersion() {
        return currentJavaVersion;
    }

    public String getTargetJavaURL() {
        return targetJavaURL;
    }

    public String getTempDirectory() {
        if (tempDirectory == null) {
            throw new IceBoarException("User temp directory is not defined!", null);
        }
        return tempDirectory;
    }

    public String getDestinationJREPath() {
        String fileName = targetJavaURL.substring(targetJavaURL.lastIndexOf('/') + 1);
        return getTempDirectory() + fileName;
    }

    public String getDestinationJARsPath() {
        return getTempDirectory() + "IceBoar_" + jvmStartTime;
    }

    public String getUnzipPath() {
        String tempDirectory = getTempDirectory();
        String fileNameWithoutExtension = targetJavaURL
                .substring(targetJavaURL.lastIndexOf('/') + 1, targetJavaURL.lastIndexOf('.'));
        return tempDirectory + fileNameWithoutExtension + "_" + jvmStartTime;
    }

    public String getUnzipJavaCommandPath() {
        return FileUtilsFacade.addJavaCommandPathToPath(getUnzipPath());
    }

    public List<String> getApplicationArguments() {
        return applicationArguments;
    }

    public String getApplicationArgumentsAsString() {
        return join(applicationArguments, " ");
    }

    public List<String> getJarURLs() {
        return jarURLs;
    }


    public String getMainClass() {
        return mainClass;
    }

    public String getClassPathAsText() {
        List<String> jars = new ArrayList<String>();
        for (String jarURL : jarURLs) {
            jars.add(getDestinationPathForJar(jarURL));
        }
        return join(jars, getPathSeparator());
    }

    public String getDestinationPathForJar(String jarUrl) {
        return getDestinationJARsPath() + File.separator + extractFilenameFromURL(jarUrl);
    }

    public List<String> getAllPropertiesForTarget() {
        return allPropertiesForTarget;
    }

    public String getAllPropertiesForTargetAsText() {
        return join(allPropertiesForTarget, " ");
    }

    public String getOperationSystemName() {
        return operationSystemName;
    }

    public boolean isOperationSystemMacOSX() {
        return getOperationSystemName().equals("Mac OS X");
    }

    public String getPathSeparator() {
        return pathSeparator;
    }

    public String getInitialHeapSize() {
        if (isNotBlank(initialHeapSize)) {
            return "-Xms" + initialHeapSize;
        }
        return "";
    }

    public String getMaxHeapSize() {
        if (isNotBlank(maxHeapSize)) {
            return "-Xmx" + maxHeapSize;
        }
        return "";
    }

    public List<String> getJavaVmArgs() {
        if (javaVmArgs == null || javaVmArgs.isEmpty()) {
            return new ArrayList<String>();
        }
        List<String> result = Arrays.asList(javaVmArgs.split(" "));
        return result;
    }

    public String getJavaVmArgsAsText() {
        if (isNotBlank(javaVmArgs)) {
            return javaVmArgs;
        }
        return "";
    }

    public String getCachePath() {
        return cachePath;
    }

    public CacheStatus getCacheStatus() {
        return cacheStatus;
    }

    public String getDestinationJreZipPathFromCache() {
        CacheStatus cacheStatus = getCacheStatus();
        StatusInfo statusInfo = cacheStatus.getJreDownloadedStatusInfo(getTargetJavaVersion());
        if (statusInfo == null || statusInfo.getPath() == null) {
            return "";
        }
        return statusInfo.getPath();
    }

    public String getUnzippedJrePathFromCache() {
        CacheStatus cacheStatus = getCacheStatus();
        StatusInfo statusInfo = cacheStatus.getJreUnzippedStatusInfo(getTargetJavaVersion());
        if (statusInfo == null || statusInfo.getPath() == null) {
            return "";
        }
        return statusInfo.getPath();
    }

    public boolean isCloseOnEnd() {
        return closeOnEnd;
    }

    public List<String> getIcons() {
        return icons;
    }

    public String getCustomSplashImage() {
        return customSplashImage;
    }

    public boolean isHideFrameBorder() {
        return hideFrameBorder;
    }

    public boolean runOnTargetJVM() {
        return !versionMatcher.match(this) || isAlwaysRunOnTargetJVM();
    }

    private boolean isAlwaysRunOnTargetJVM() {
        return alwaysRunOnTargetJVM;
    }

    public String getCurrentJavaCommand() {
        return removeQuotationMarksAtBeginAndEnd(currentJavaCommand);
    }

    private String removeQuotationMarksAtBeginAndEnd(String input) {
        String text = input;
        if(input.charAt(0) == '\"') {
            text = text.substring(1, text.length());
        }
        if(input.charAt(input.length() - 1) == '\"') {
            text = text.substring(0, text.length() - 1);
        }
        return text;
    }

    public static class Builder {
        private String[] applicationArguments = new String[]{};
        private long jvmStartTime;
        private boolean showDebug;
        private String frameTitle;
        private String targetJavaVersion;
        private String currentJavaVersion;
        private String targetJavaURL;
        private String tempDirectory;
        private List<String> jarURLs;
        private String mainClass;
        private List<String> allPropertiesForTarget;
        private String operationSystemName;
        private String pathSeparator;
        private String initialHeapSize;
        private String maxHeapSize;
        private String javaVmArgs;
        private String cachePath;
        private CacheStatus cacheStatus;
        private boolean closeOnEnd;
        private List<String> icons;
        private String splashScreen;
        private boolean hideFrameBorder;
        private boolean alwaysRunOnTargetJVM;
        private String currentJavaCommand;

        public Builder applicationArguments(String[] applicationArguments) {
            if (applicationArguments != null) {
                this.applicationArguments = applicationArguments.clone();
            }
            return this;
        }

        public Builder jvmStartTime(long jvmStartTime) {
            this.jvmStartTime = jvmStartTime;
            return this;
        }

        public Builder frameTitle(String frameTitle) {
            this.frameTitle = frameTitle;
            return this;
        }

        public Builder showDebug(boolean showDebug) {
            this.showDebug = showDebug;
            return this;
        }

        public Builder targetJavaVersion(String targetJavaVersion) {
            this.targetJavaVersion = targetJavaVersion;
            return this;
        }

        public Builder currentJavaVersion(String currentJavaVersion) {
            this.currentJavaVersion = currentJavaVersion;
            return this;
        }

        public Builder targetJavaURL(String targetJavaURL) {
            this.targetJavaURL = targetJavaURL;
            return this;
        }

        public Builder tempDirectory(String tempDirectory) {
            if (tempDirectory.endsWith(File.separator)) {
                this.tempDirectory = tempDirectory;
            } else {
                this.tempDirectory = tempDirectory + File.separator;
            }
            return this;
        }

        public Builder jarURLs(List<String> jarURLs) {
            this.jarURLs = jarURLs;
            return this;
        }

        public Builder mainClass(String mainClass) {
            this.mainClass = mainClass;
            return this;
        }

        public Builder allPropertiesForTarget(List<String> allPropertiesForTarget) {
            this.allPropertiesForTarget = allPropertiesForTarget;
            return this;
        }

        public Builder operationSystemName(String operationSystemName) {
            this.operationSystemName = operationSystemName;
            return this;
        }

        public Builder pathSeparator(String pathSeparator) {
            this.pathSeparator = pathSeparator;
            return this;
        }

        public Builder initialHeapSize(String initialHeapSize) {
            this.initialHeapSize = initialHeapSize;
            return this;
        }

        public Builder maxHeapSize(String maxHeapSize) {
            this.maxHeapSize = maxHeapSize;
            return this;
        }

        public Builder javaVmArgs(String javaVmArgs) {
            this.javaVmArgs = javaVmArgs;
            return this;
        }

        public Builder closeOnEnd(boolean closeOnEnd) {
            this.closeOnEnd = closeOnEnd;
            return this;
        }

        public Builder cachePath(String cachePath) {
            this.cachePath = cachePath;
            return this;
        }

        public Builder cacheStatus(CacheStatus cacheStatus) {
            this.cacheStatus = cacheStatus;
            return this;
        }

        public Builder icons(List<String> icons) {
            this.icons = icons;
            return this;
        }

        public Builder customSplashImage(String image) {
            this.splashScreen = image;
            return this;
        }

        public Builder hideFrameBorder(boolean hideFrameBorder) {
            this.hideFrameBorder = hideFrameBorder;
            return this;
        }

        public Builder alwaysRunOnTargetJVM(boolean alwaysRunOnTargetJVM) {
            this.alwaysRunOnTargetJVM = alwaysRunOnTargetJVM;
            return this;
        }

        public Builder currentJavaCommand(String currentJvmPath) {
            this.currentJavaCommand = currentJvmPath;
            return this;
        }

        public GlobalSettings build() {
            GlobalSettings settings = new GlobalSettings(applicationArguments);
            settings.jvmStartTime = jvmStartTime;
            settings.showDebug = showDebug;
            settings.frameTitle = frameTitle;
            settings.targetJavaVersion = targetJavaVersion;
            settings.currentJavaVersion = currentJavaVersion;
            settings.targetJavaURL = targetJavaURL;
            settings.tempDirectory = tempDirectory;
            settings.jarURLs = jarURLs;
            settings.mainClass = mainClass;
            settings.allPropertiesForTarget = allPropertiesForTarget;
            settings.operationSystemName = operationSystemName;
            settings.pathSeparator = pathSeparator;
            settings.initialHeapSize = initialHeapSize;
            settings.maxHeapSize = maxHeapSize;
            settings.javaVmArgs = javaVmArgs;
            settings.closeOnEnd = closeOnEnd;
            settings.cachePath = cachePath;
            settings.cacheStatus = cacheStatus;
            settings.icons = icons;
            settings.customSplashImage = splashScreen;
            settings.hideFrameBorder = hideFrameBorder;
            settings.alwaysRunOnTargetJVM = alwaysRunOnTargetJVM;
            settings.currentJavaCommand = currentJavaCommand;
            return settings;
        }
    }
}
